# Fifa practice by Crist Alcázar

We want to predict the *value* of the players using the information of the player.

The files of this project are in this <a href="https://github.com/kralcazar/practica_inteligencia">repository</a>.

## Import libraries

First of all we import the libraries we're gonna use.

In [1]:
import os
import time
import datetime

from sklearn.model_selection import train_test_split
from sklearn import linear_model
import matplotlib.pyplot as plt
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

import pandas as pd
import numpy as np

## Read the data

To read the data we're gonna use ``pandas`` and an utility from the library
``os``.


In [2]:
df = pd.read_csv(os.path.join("..", "in", "fifa.csv"))
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club  ... Composure Marking StandingTackle  SlidingTackle  \
0         FC Barcelona  ...      96.0    33.0           28.0           26.0   
1             Juventus  ...      95.0    28.0           31.0           23.0   
2  Paris Saint-Germain  ...      94.0    27.0           24.0           33.0   
3    Manchester United  ...      68.0    15.0           21.0           13.0   
4      Manchester City  ...      88.0    68.0           58.0           51.0   

  GKDiving  GKHandling  GKKicking  GKPositioning GKReflexes Release Clause  
0      6.0        11.0       15.0           14.0        8.0        €226.5M  
1      7.0        11.0       15.0           14.0       11.0        €127.1M  
2      9.0         9.0       15.0           15.0       11.0        €228.1M  
3     90.0        85.0       87.0           88.0       94.0        €138.6M  
4     15.0        13.0        5.0           10.0       13.0        €196.4M  

[5 rows x 89 columns]

## Data analysis 

First of all we want to known what is happening to the data. To do so we increase the number of columns to show. After that we're gonna use a special instruction that will show us the data structure.



In [3]:
pd.set_option('display.max_columns', None)
df.head()

Unnamed: 0      ID               Name  Age  \
0           0  158023           L. Messi   31   
1           1   20801  Cristiano Ronaldo   33   
2           2  190871          Neymar Jr   26   
3           3  193080             De Gea   27   
4           4  192985       K. De Bruyne   27   

                                            Photo Nationality  \
0  https://cdn.sofifa.org/players/4/19/158023.png   Argentina   
1   https://cdn.sofifa.org/players/4/19/20801.png    Portugal   
2  https://cdn.sofifa.org/players/4/19/190871.png      Brazil   
3  https://cdn.sofifa.org/players/4/19/193080.png       Spain   
4  https://cdn.sofifa.org/players/4/19/192985.png     Belgium   

                                  Flag  Overall  Potential  \
0  https://cdn.sofifa.org/flags/52.png       94         94   
1  https://cdn.sofifa.org/flags/38.png       94         94   
2  https://cdn.sofifa.org/flags/54.png       92         93   
3  https://cdn.sofifa.org/flags/45.png       91         93   
4   https://cdn.sofifa.org/flags/7.png       91         92   

                  Club                                     Club Logo    Value  \
0         FC Barcelona  https://cdn.sofifa.org/teams/2/light/241.png  €110.5M   
1             Juventus   https://cdn.sofifa.org/teams/2/light/45.png     €77M   
2  Paris Saint-Germain   https://cdn.sofifa.org/teams/2/light/73.png  €118.5M   
3    Manchester United   https://cdn.sofifa.org/teams/2/light/11.png     €72M   
4      Manchester City   https://cdn.sofifa.org/teams/2/light/10.png    €102M   

    Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0  €565K     2202           Left                       5.0        4.0   
1  €405K     2228          Right                       5.0        4.0   
2  €290K     2143          Right                       5.0        5.0   
3  €260K     1471          Right                       4.0        3.0   
4  €355K     2281          Right                       4.0        5.0   

   Skill Moves       Work Rate   Body Type Real Face Position  Jersey Number  \
0          4.0  Medium/ Medium       Messi       Yes       RF           10.0   
1          5.0       High/ Low  C. Ronaldo       Yes       ST            7.0   
2          5.0    High/ Medium      Neymar       Yes       LW           10.0   
3          1.0  Medium/ Medium        Lean       Yes       GK            1.0   
4          4.0      High/ High      Normal       Yes      RCM            7.0   

         Joined Loaned From Contract Valid Until Height  Weight    LS    ST  \
0   Jul 1, 2004         NaN                 2021    5'7  159lbs  88+2  88+2   
1  Jul 10, 2018         NaN                 2022    6'2  183lbs  91+3  91+3   
2   Aug 3, 2017         NaN                 2022    5'9  150lbs  84+3  84+3   
3   Jul 1, 2011         NaN                 2020    6'4  168lbs   NaN   NaN   
4  Aug 30, 2015         NaN                 2023   5'11  154lbs  82+3  82+3   

     RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM   LCM    CM  \
0  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2   
1  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3   
2  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3  87+3  87+3   

    RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB   RCB    RB  \
0  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2  47+2  59+2   
1  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3  53+3  61+3   
2  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3  47+3  60+3   
3   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
4  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3  66+3  73+3   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     8

## Cleaning non relevant columns

We can see that the two first columns do not provide any meaningful information so we can delete them. We will also clean all columns that, after being proved, do not contains relevant information because them worsen the score. The keeped columns have proved that benefits the final score.

In [4]:
df = df.iloc[:, 2:]
df.drop(['Name','Photo','Flag','Club Logo','Real Face','Body Type','Jersey Number'], axis=1, inplace=True)
df

Age Nationality  Overall  Potential                 Club    Value  \
0       31   Argentina       94         94         FC Barcelona  €110.5M   
1       33    Portugal       94         94             Juventus     €77M   
2       26      Brazil       92         93  Paris Saint-Germain  €118.5M   
3       27       Spain       91         93    Manchester United     €72M   
4       27     Belgium       91         92      Manchester City    €102M   
...    ...         ...      ...        ...                  ...      ...   
18202   19     England       47         65      Crewe Alexandra     €60K   
18203   19      Sweden       47         63       Trelleborgs FF     €60K   
18204   16     England       47         67     Cambridge United     €60K   
18205   17     England       47         66      Tranmere Rovers     €60K   
18206   16     England       46         66      Tranmere Rovers     €60K   

        Wage  Special Preferred Foot  International Reputation  Weak Foot  \
0      €565K     2202           Left                       5.0        4.0   
1      €405K     2228          Right                       5.0        4.0   
2      €290K     2143          Right                       5.0        5.0   
3      €260K     1471          Right                       4.0        3.0   
4      €355K     2281          Right                       4.0        5.0   
...      ...      ...            ...                       ...        ...   
18202    €1K     1307          Right                       1.0        2.0   
18203    €1K     1098          Right                       1.0        2.0   
18204    €1K     1189          Right                       1.0        3.0   
18205    €1K     1228          Right                       1.0        3.0   
18206    €1K     1321          Right                       1.0        3.0   

       Skill Moves       Work Rate Position        Joined Loaned From  \
0              4.0  Medium/ Medium       RF   Jul 1, 2004         NaN   
1              5.0       High/ Low       ST  Jul 10, 2018         NaN   
2              5.0    High/ Medium       LW   Aug 3, 2017         NaN   
3              1.0  Medium/ Medium       GK   Jul 1, 2011         NaN   
4              4.0      High/ High      RCM  Aug 30, 2015         NaN   
...            ...             ...      ...           ...         ...   
18202          2.0  Medium/ Medium       CM   May 3, 2017         NaN   
18203          2.0  Medium/ Medium       ST  Mar 19, 2018         NaN   
18204          2.0  Medium/ Medium       ST   Jul 1, 2017         NaN   
18205          2.0  Medium/ Medium       RW  Apr 24, 2018         NaN   
18206          2.0  Medium/ Medium       CM  Oct 30, 2018         NaN   

      Contract Valid Until Height  Weight    LS    ST    RS    LW    LF    CF  \
0                     2021    5'7  159lbs  88+2  88+2  88+2  92+2  93+2  93+2   
1                     2022    6'2  183lbs  91+3  91+3  91+3  89+3  90+3  90+3   
2                     2022    5'9  150lbs  84+3  84+3  84+3  89+3  89+3  89+3   
3                     2020    6'4  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   
4                     2023   5'11  154lbs  82+3  82+3  82+3  87+3  87+3  87+3   
...                    ...    ...     ...   ...   ...   ...   ...   ...   ...   
18202                 2019    5'9  134lbs  42+2  42+2  42+2  44+2  44+2  44+2   
18203                 2020    6'3  170lbs  45+2  45+2  45+2  39+2  42+2  42+2   
18204                 2021    5'8  148lbs  45+2  45+2  45+2  45+2  46+2  46+2   
18205                 2019   5'10  154lbs  47+2  47+2  47+2  47+2  46+2  46+2   
18206                 2019   5'10  176lbs  43+2  43+2  43+2  45+2  44+2  44+2   

         RF    RW   LAM   CAM   RAM    LM   LCM    CM   RCM    RM   LWB   LDM  \
0      93+2  92+2  93+2  93+2  93+2  91+2  84+2  84+2  84+2  91+2  64+2  61+2   
1      90+3  89+3  88+3  88+3  88+3  88+3  81+3  81+3  81+3  88+3  65+3  61+3   
2      89+3  89+3  89+3  89+3  89+3  88+3  81+3  81+3  81+3  88+3  65+3  60+

The data contains *NaN* (Not a Number). What it mean a *NaN*? What we do with *NaN*

## Handle NaNs

Let's see which columns contain NaNs.

In [5]:
df.columns[df.isna().any()].tolist()

['Club',
 'Preferred Foot',
 'International Reputation',
 'Weak Foot',
 'Skill Moves',
 'Work Rate',
 'Position',
 'Joined',
 'Loaned From',
 'Contract Valid Until',
 'Height',
 'Weight',
 'LS',
 'ST',
 'RS',
 'LW',
 'LF',
 'CF',
 'RF',
 'RW',
 'LAM',
 'CAM',
 'RAM',
 'LM',
 'LCM',
 'CM',
 'RCM',
 'RM',
 'LWB',
 'LDM',
 'CDM',
 'RDM',
 'RWB',
 'LB',
 'LCB',
 'CB',
 'RCB',
 'RB',
 'Crossing',
 'Finishing',
 'HeadingAccuracy',
 'ShortPassing',
 'Volleys',
 'Dribbling',
 'Curve',
 'FKAccuracy',
 'LongPassing',
 'BallControl',
 'Acceleration',
 'SprintSpeed',
 'Agility',
 'Reactions',
 'Balance',
 'ShotPower',
 'Jumping',
 'Stamina',
 'Strength',
 'LongShots',
 'Aggression',
 'Interceptions',
 'Positioning',
 'Vision',
 'Penalties',
 'Composure',
 'Marking',
 'StandingTackle',
 'SlidingTackle',
 'GKDiving',
 'GKHandling',
 'GKKicking',
 'GKPositioning',
 'GKReflexes',
 'Release Clause']

We can see that almost all players has a NaN. We can see multiple columns that always are a NaN as for example **Loaned From**. And some problematic player as David DeGea. And the rest of columns that we known that haves NaN as for example the **Club** columns what can we do?.

We can get all the rows with the column **Club** equals to NaN. 

In [6]:
df.loc[df.Club != df.Club]

Age Nationality  Overall  Potential Club Value Wage  Special  \
452     24   Argentina       80         85  NaN    €0   €0     2122   
538     33      Sweden       80         80  NaN    €0   €0     1797   
568     26      Russia       79         81  NaN    €0   €0     1217   
677     29      Russia       79         79  NaN    €0   €0     2038   
874     29      Russia       78         78  NaN    €0   €0     1810   
...    ...         ...      ...        ...  ...   ...  ...      ...   
17197   21       India       55         64  NaN    €0   €0      838   
17215   26     Finland       55         57  NaN    €0   €0     1366   
17339   23       India       54         63  NaN    €0   €0     1321   
17436   20       India       54         67  NaN    €0   €0     1270   
17539   21       India       53         62  NaN    €0   €0     1247   

      Preferred Foot  International Reputation  Weak Foot  Skill Moves  \
452            Right                       2.0        4.0          4.0   
538            Right                       2.0        4.0          2.0   
568            Right                       1.0        3.0          1.0   
677            Right                       2.0        3.0          3.0   
874            Right                       2.0        3.0          3.0   
...              ...                       ...        ...          ...   
17197          Right                       1.0        2.0          1.0   
17215          Right                       1.0        3.0          2.0   
17339          Right                       1.0        3.0          2.0   
17436          Right                       1.0        3.0          2.0   
17539          Right                       1.0        3.0          2.0   

            Work Rate Position Joined Loaned From Contract Valid Until Height  \
452    Medium/ Medium       CM    NaN         NaN                  NaN   5'11   
538      High/ Medium      LCB    NaN         NaN                  NaN    6'4   
568    Medium/ Medium       GK    NaN         NaN                  NaN    6'2   
677        High/ High       RB    NaN         NaN                  NaN   5'10   
874      High/ Medium       ST    NaN         NaN                  NaN    6'5   
...               ...      ...    ...         ...                  ...    ...   
17197  Medium/ Medium       GK    NaN         NaN                  NaN    6'2   
17215    Medium/ High       RB    NaN         NaN                  NaN    6'4   
17339     Medium/ Low      NaN    NaN         NaN                  NaN    5'9   
17436  Medium/ Medium      NaN    NaN         NaN                  NaN    6'0   
17539    High/ Medium      NaN    NaN         NaN                  NaN    6'3   

       Weight    LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM  \
452    165lbs  71+2  71+2  71+2  75+2  75+2  75+2  75+2  75+2  77+2  77+2   
538    185lbs  62+2  62+2  62+2  56+2  58+2  58+2  58+2  56+2  58+2  58+2   
568    176lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677    154lbs  70+2  70+2  70+2  73+2  72+2  72+2  72+2  73+2  73+2  73+2   
874    201lbs  77+2  77+2  77+2  71+2  74+2  74+2  74+2  71+2  71+2  71+2   
...       ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   ...   
17197  176lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17215  187lbs  46+2  46+2  46+2  45+2  44+2  44+2  44+2  45+2  43+2  43+2   
17339  143lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17436  168lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
17539  174lbs   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   

        RAM    LM   LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB  \
452    77+2  76+2  79+2  79+2  79+2  76+2  75+2  77+2  77+2  77+2  75+2  74+2   
538    58+2  57+2  64+2  64+2  64+2  57+2  68+2  74+2  74+2  74+2  68+2  70+2   
568     NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
677    73+2  75+2  74+2  74+2  74+2  75+2  

The problem of these players is that their club is not on FIFA.

So we can try to delete all the players with the club with *NaN*. To do so we can use the next instruction. We will also clean players without known position neither the goal keepers.

In [7]:
# It's difficult to get the player's value without knowing their club
# These players also don't have information for ['Joined', 'Loaned From', 'Contract Valid Until', 'Release Clause']
df = df.dropna(subset = ['Club'])
# Dropping the players who have NaN in position cleans a lot of NaNs scores like LS, ST, etc.
# It's difficult to get the player's value without knowing their positions
df = df.dropna(subset = ['Position'])
# For some reason the goal keeper information worsens the score, it also cleans a lot of NaN data
df = df.drop(df[df['Position']=='GK'].index)

## Distribute non numerical columns

We can see that we removed all the players without a known club. Nevertheless we still have a problem, how to use this feature for an algorithm?

So we distribute the columns that do not have numerical values to become a unique column with value 0 if they do not have this value and 1 value if they do.

In [8]:
clb = df.pop("Club")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(clb, prefix='clb').reset_index(drop=True)], axis=1, sort=False)

prf = df.pop("Preferred Foot")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(prf, prefix='prf').reset_index(drop=True)], axis=1, sort=False)

wr = df.pop("Work Rate")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(wr, prefix='wr').reset_index(drop=True)], axis=1, sort=False)

pos = df.pop("Position")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(pos, prefix='pos').reset_index(drop=True)], axis=1, sort=False)

nat = df.pop("Nationality")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(nat, prefix='nat').reset_index(drop=True)], axis=1, sort=False)

lf = df.pop("Loaned From")
df = pd.concat([df.reset_index(drop=True), pd.get_dummies(lf, prefix='lf').reset_index(drop=True)], axis=1, sort=False)

df.head()

Age  Overall  Potential    Value   Wage  Special  International Reputation  \
0   31       94         94  €110.5M  €565K     2202                       5.0   
1   33       94         94     €77M  €405K     2228                       5.0   
2   26       92         93  €118.5M  €290K     2143                       5.0   
3   27       91         92    €102M  €355K     2281                       4.0   
4   27       91         91     €93M  €340K     2142                       4.0   

   Weak Foot  Skill Moves        Joined Contract Valid Until Height  Weight  \
0        4.0          4.0   Jul 1, 2004                 2021    5'7  159lbs   
1        4.0          5.0  Jul 10, 2018                 2022    6'2  183lbs   
2        5.0          5.0   Aug 3, 2017                 2022    5'9  150lbs   
3        5.0          4.0  Aug 30, 2015                 2023   5'11  154lbs   
4        4.0          4.0   Jul 1, 2012                 2020    5'8  163lbs   

     LS    ST    RS    LW    LF    CF    RF    RW   LAM   CAM   RAM    LM  \
0  88+2  88+2  88+2  92+2  93+2  93+2  93+2  92+2  93+2  93+2  93+2  91+2   
1  91+3  91+3  91+3  89+3  90+3  90+3  90+3  89+3  88+3  88+3  88+3  88+3   
2  84+3  84+3  84+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  89+3  88+3   
3  82+3  82+3  82+3  87+3  87+3  87+3  87+3  87+3  88+3  88+3  88+3  88+3   
4  83+3  83+3  83+3  89+3  88+3  88+3  88+3  89+3  89+3  89+3  89+3  89+3   

    LCM    CM   RCM    RM   LWB   LDM   CDM   RDM   RWB    LB   LCB    CB  \
0  84+2  84+2  84+2  91+2  64+2  61+2  61+2  61+2  64+2  59+2  47+2  47+2   
1  81+3  81+3  81+3  88+3  65+3  61+3  61+3  61+3  65+3  61+3  53+3  53+3   
2  81+3  81+3  81+3  88+3  65+3  60+3  60+3  60+3  65+3  60+3  47+3  47+3   
3  87+3  87+3  87+3  88+3  77+3  77+3  77+3  77+3  77+3  73+3  66+3  66+3   
4  82+3  82+3  82+3  89+3  66+3  63+3  63+3  63+3  66+3  60+3  49+3  49+3   

    RCB    RB  Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  \
0  47+2  59+2      84.0       95.0             70.0          90.0     86.0   
1  53+3  61+3      84.0       94.0             89.0          81.0     87.0   
2  47+3  60+3      79.0       87.0             62.0          84.0     84.0   
3  66+3  73+3      93.0       82.0             55.0          92.0     82.0   
4  49+3  60+3      81.0       84.0             61.0          89.0     80.0   

   Dribbling  Curve  FKAccuracy  LongPassing  BallControl  Acceleration  \
0       97.0   93.0        94.0         87.0         96.0          91.0   
1       88.0   81.0        76.0         77.0         94.0          89.0   
2       96.0   88.0        87.0         78.0         95.0          94.0   
3       86.0   85.0        83.0         91.0         91.0          78.0   
4       95.0   83.0        79.0         83.0         94.0          94.0   

   SprintSpeed  Agility  Reactions  Balance  ShotPower  Jumping  Stamina  \
0         86.0     91.0       95.0     95.0       85.0     68.0     72.0   
1         91.0     87.0       96.0     70.0       95.0     95.0     88.0   
2         90.0     96.0       94.0     84.0       80.0     61.0     81.0   
3         76.0     79.0       91.0     77.0       91.0     63.0     90.0   
4         88.0     95.0       90.0     94.0       82.0     56.0     83.0   

   Strength  LongShots  Aggression  Interceptions  Positioning  Vision  \
0      59.0       94.0        48.0           22.0         94.0    94.0   
1      79.0       93.0        63.0           29.0         95.0    82.0   
2      49.0       82.0        56.0           36.0         89.0    87.0   
3      75.0       91.0        76.0           61.0         87.0    94.0   
4      66.0       80.0        54.0           41.0         87.0    89.0   

   Penalties  Composure  Marking  StandingTackle  SlidingTackle  GKDiving  \
0       75.0       96.0     33.0            28.0           26.0       6.0   
1       85.0       95.0     28.0            31.0           23.0       7.0   
2       81.0       94.0     27.0            24.0           33.

## Units and numerical formating

Now we fixed the non numerical columns but what about the **Values**, it does not have any NaN but it is not a number. How we convert the format?

First of all we define all the functions to pass from one format to the other:

In [9]:
def value_to_int(x):
    # From K and M to int.

    ret_val = 0
    
    # For non assigned clause release we will return 0 assuming that these players do not have release clause to leave the team.
    if pd.isna(x) or x is np.nan or pd.isnull(x):
        return ret_val
    
    # Remove € unit
    x = x.replace('€', '')    
    
    # Removing and normalizing thousands and millions "shortcuts"
    if type(x) == float or type(x) == int:
        ret_val = x
    if 'K' in x:
        if len(x) > 1:
            ret_val = float(x.replace('K', ''))
        ret_val = ret_val *1000
    if 'M' in x:
        if len(x) > 1:
            ret_val = float(x.replace('M', ''))
        ret_val = ret_val * 1000000
    return int(ret_val)

def remove_lbs(x):
    # Remove lbs unit
    x = int(x.replace('lbs', ''))
    return x

def sum_to_int(serie):
    # Apply the sum in values like (89+2)
    for index, value in serie.items():
        a = value.partition("+")[0]
        b = value.partition("+")[2]
        serie[index] = int(a) + int(b)
    return serie

def format_apostrophes_to_float(x):
    # Convert the british pound comma style to float
    return float(x.replace("'", "."))

def correct_nan_joined(age, joined):
    # Apply one single date for the players that do not have an assigned Joined date
    if pd.isna(joined) or joined is np.nan or pd.isnull(joined):
        # We assume that the players were being in club since they were young 
        # 6 year old is the mean age for the players were joining a club
        # Other tested ages return a worse score
        return str(2020-int(age)+6)
    return joined

def stringDate_to_timestamp(str):
    # From date in string to timestamp numerical int
    value=0
    if len(str)==4: # For the single year date (2014)
        value= time.mktime(datetime.datetime.strptime(str, "%Y").timetuple())
    else: # For the full date (Jul 1, 2004)
        value=time.mktime(datetime.datetime.strptime(str, "%b %d, %Y").timetuple())
    return int(value)

Now we want to apply these functions to each value of the needed columns so we can train the prediction with them.

In [10]:
# Apply for thouse that do not have Joined date
df['Joined'] = df.apply(lambda x: correct_nan_joined(x['Age'], x['Joined']), axis=1)
# Convert the string Joined dates to numerical dates
df["Joined"] = df["Joined"].apply(stringDate_to_timestamp)
# Convert the string Contract Valid Unitl dates to numerical dates
df["Contract Valid Until"] = df["Contract Valid Until"].apply(stringDate_to_timestamp)

# Format values to int with an unit in the value
df["Value"] = df["Value"].apply(value_to_int)
df["Wage"] = df["Wage"].apply(value_to_int)
df["Release Clause"] = df["Release Clause"].apply(value_to_int)
df["Weight"] = df["Weight"].apply(remove_lbs)
df["Height"] = df["Height"].apply(format_apostrophes_to_float)

# Apply the sum embedded in value
df[['LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB']] = df[['LS','ST','RS','LW','LF','CF','RF','RW','LAM','CAM','RAM','LM','LCM','CM','RCM','RM','LWB','LDM','CDM','RDM','RWB','LB','LCB','CB','RCB','RB']].apply(sum_to_int)

# Show the results
df.head()

Age  Overall  Potential      Value    Wage  Special  \
0   31       94         94  110500000  565000     2202   
1   33       94         94   77000000  405000     2228   
2   26       92         93  118500000  290000     2143   
3   27       91         92  102000000  355000     2281   
4   27       91         91   93000000  340000     2142   

   International Reputation  Weak Foot  Skill Moves      Joined  \
0                       5.0        4.0          4.0  1088632800   
1                       5.0        4.0          5.0  1531173600   
2                       5.0        5.0          5.0  1501711200   
3                       4.0        5.0          4.0  1440885600   
4                       4.0        4.0          4.0  1341093600   

   Contract Valid Until  Height  Weight  LS  ST  RS  LW  LF  CF  RF  RW LAM  \
0            1609455600    5.70     159  90  90  90  94  95  95  95  94  95   
1            1640991600    6.20     183  94  94  94  92  93  93  93  92  91   
2            1640991600    5.90     150  87  87  87  92  92  92  92  92  92   
3            1672527600    5.11     154  85  85  85  90  90  90  90  90  91   
4            1577833200    5.80     163  86  86  86  92  91  91  91  92  92   

  CAM RAM  LM LCM  CM RCM  RM LWB LDM CDM RDM RWB  LB LCB  CB RCB  RB  \
0  95  95  93  86  86  86  93  66  63  63  63  66  61  49  49  49  61   
1  91  91  91  84  84  84  91  68  64  64  64  68  64  56  56  56  64   
2  92  92  91  84  84  84  91  68  63  63  63  68  63  50  50  50  63   
3  91  91  91  90  90  90  91  80  80  80  80  80  76  69  69  69  76   
4  92  92  92  85  85  85  92  69  66  66  66  69  63  52  52  52  63   

   Crossing  Finishing  HeadingAccuracy  ShortPassing  Volleys  Dribbling  \
0      84.0       95.0             70.0          90.0     86.0       97.0   
1      84.0       94.0             89.0          81.0     87.0       88.0   
2      79.0       87.0             62.0          84.0     84.0       96.0   
3      93.0       82.0             55.0          92.0     82.0       86.0   
4      81.0       84.0             61.0          89.0     80.0       95.0   

   Curve  FKAccuracy  LongPassing  BallControl  Acceleration  SprintSpeed  \
0   93.0        94.0         87.0         96.0          91.0         86.0   
1   81.0        76.0         77.0         94.0          89.0         91.0   
2   88.0        87.0         78.0         95.0          94.0         90.0   
3   85.0        83.0         91.0         91.0          78.0         76.0   
4   83.0        79.0         83.0         94.0          94.0         88.0   

   Agility  Reactions  Balance  ShotPower  Jumping  Stamina  Strength  \
0     91.0       95.0     95.0       85.0     68.0     72.0      59.0   
1     87.0       96.0     70.0       95.0     95.0     88.0      79.0   
2     96.0       94.0     84.0       80.0     61.0     81.0      49.0   
3     79.0       91.0     77.0       91.0     63.0     90.0      75.0   
4     95.0       90.0     94.0       82.0     56.0     83.0      66.0   

   LongShots  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       94.0        48.0           22.0         94.0    94.0       75.0   
1       93.0        63.0           29.0         95.0    82.0       85.0   
2       82.0        56.0           36.0         89.0    87.0       81.0   
3       91.0        76.0           61.0         87.0    94.0       79.0   
4       80.0        54.0           41.0         87.0    89.0       86.0   

   Composure  Marking  StandingTackle  SlidingTackle  GKDiving  GKHandling  \
0       96.0     33.0            28.0           26.0       6.0        11.0   
1       95.0     28.0            31.0           23.0       7.0        11.0   
2       94.0     27.0            24.0           33.0       9.0         9.0   
3       88.0     68.0            58.0           51.0      15.0        13.0   
4       91.0     34.0            27.0           22.0      11.0        12.0   

   GKKicking  GKPositioning  GKReflexes  Release Cla

Now we have all columns with numerical and trainable values.

## Prediction

Finally we want to try to predict the **value**, to do so we store this column in the val variable.

In [11]:
val = df.pop("Value")

Now we have the data we want to predict in a numeric format, the value information. So now we can use this data to train a model. But first we need to split the data to being able to known its true performance.

We assign a test size value, which it will be the non trained data for testing the results after the training. The recomendation in to set it at 20%.

In [12]:
X_train, X_test, y_train, y_test = train_test_split(df, val, test_size=0.20, random_state=42)

In [13]:
len(X_train)

12740

We have 12740 rows to train the model. 

We will use a LinearRegression model because we need to obtain a continous value from multidimensional dataset and can support some dominant classes and small variations in data without change its prediction too much. 

The higher cost of the computing time is not an issue for this kind of training so the LinearRegression model for this problem is a good choice.

Let's do the train fitting the splitted data.

In [14]:
reg = linear_model.LinearRegression().fit(X_train, y_train)

Predict the Value using the linear model of data testing splited.

In [15]:
preds = reg.predict(X_test)

Let's see the first row **predicted** value.

In [16]:
preds[0]

738540.7504741431

Let's see the first row **real** value.

In [17]:
y_test[0]

110500000

Let's see the coeficients.

In [18]:
# The coefficients
print('Coefficients: \n', reg.coef_)
# The mean squared error
print('Mean squared error: %.2f' % mean_squared_error(y_test, preds))
# The coefficient of determination: 1 is perfect prediction
print('Coefficient of determination: %.2f' % r2_score(y_test, preds))

Coefficients: 
 [-56276.26181173  71126.96011495 -15303.90321458 ...  20011.34688828
      0.         -56102.01678503]
Mean squared error: 850784715073.95
Coefficient of determination: 0.98


Finally we get a metric $R^2$ for the regression, we use the implementation from <a href="https://scikit-learn.org/stable/modules/generated/sklearn.metrics.r2_score.html"> sickit-learn</a>.

In [19]:
r2_score(preds, y_test)

0.9785421951971136

We obtanined a good score (97,85%)